### PR TITLE
[compiler] Recompute values every time

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29658
* __->__ #29657
* #29653
* #29656

Summary: This PR expands the analysis from the previous in the stack in order to also capture when a value can incorrectly change within a single render, rather than just changing between two renders. In the case where dependencies have changed and so a new value is being computed, we now compute the value twice and compare the results. This would, for example, catch when we call Math.random() in render.

The generated code is a little convoluted, because we don't want to have to traverse the generated code and substitute variable names with new ones. Instead, we save the initial value to the cache as normal, then run the computation block again and compare the resulting values to the cached ones. Then, to make sure that the cached values are identical to the computed ones, we reassign the cached values into the output variables.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts [hunk 1]
```diff
@@ -558,34 +542,10 @@ function codegenReactiveScope(
...
-            t.numericLiteral(index),
-            true
-          )
-        )
-      )
-    );
+    cacheLoads.push({ name, index, value: wrapCacheDep(cx, name) });
   }
+
+  // (changeExpressions.length > 0) || (changeExpressions.length
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/CodegenReactiveFunction.ts:548]: nit: missing closing paren here, did this get cut off maybe?

**Comment by josephsavona** [comment]
One thing to check here is what happens for scopes with reassignments rather than declarations. An example of that is:

```js
function Component(props) {
  let x = null;
  if (props.cond) {
    x = [];
    x.push(props.value);
  }
  return x;
}
```

**Comment by mvitousek** [comment]
Added! Result looks correct to me AFACT.

File: compiler/packages/snap/src/SproutTodoFilter.ts [hunk 2]
```diff
@@ -498,6 +498,7 @@ const skipFilter = new Set([
   "useState-pruned-dependency-change-detect",
   "useState-unpruned-dependency",
   "useState-and-other-hook-unpruned-dependency",
+  "change-detect-reassign",
```
Comment (josephsavona) [inline compiler/packages/snap/src/SproutTodoFilter.ts:501]: oh interesting, what was the error for this fixture?

**Review by josephsavona — APPROVED** [review]
let's land and iterate, or feel free to make final fixes without waiting for another stamp.

File: compiler/packages/snap/src/SproutTodoFilter.ts (hunk 2 above)
Comment (mvitousek) [reply compiler/packages/snap/src/SproutTodoFilter.ts:501]: It uses the react-compiler-runtime, IIUC that's not available in fixtures right?

**Pull request merged**